### PR TITLE
memtx: fix handling of corner cases gap tracking in transaction manager

### LINKED ISSUE #7375: Gap tracking corner case works incorrectly
**Steps to reproduce**

```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local ffi = require('ffi')
local log = require('log')
local txn_proxy = require('txn_proxy')
local yaml = require('yaml')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk', {parts = {{1, 'uint'}}})
box.space.s:create_index('sk', {parts = {{2, 'uint'}}})

local tx1 = txn_proxy:new()
local tx2 = txn_proxy:new()

tx1('box.begin()')
tx2('box.begin()')

tx1('box.space.s.index[1]:select({1}, {iterator = "LE", fullscan = true})')
tx1('box.space.s:insert{0, 1}')

-- Conflicts 1.
tx2('box.space.s:replace{1, 0}')
-- Conflicts 1.

tx2('box.commit()')

log.info("%s", yaml.encode(tx1('box.commit()')))

os.exit()
```

```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local ffi = require('ffi')
local json = require('json')
local log = require('log')
local txn_proxy = require('txn_proxy')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk', {parts = {{1, 'uint'}}})
box.space.s:create_index('sk', {unique = false, parts = {{2, 'uint'}}})

local tx1 = txn_proxy:new()
local tx2 = txn_proxy:new()

tx1('box.begin()')
tx2('box.begin()')

tx1('box.space.s.index[1]:select({0}, {iterator = "EQ", fullscan = true})')
tx1('box.space.s:insert{2, 0}')

-- Conflicts 1.
tx2('box.space.s:replace{1, 0}')
-- Conflicts 1.

tx2('box.commit()')

log.info("%s", json.encode(tx1('box.commit()')))

os.exit()
```

**Actual behavior**

```console
 ---
...
```

**Expected behavior**

```console
 ---
- {'error': 'Transaction has been aborted by conflict'}
...
```

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
Gap tracking does not handle gap writes when the key has the same value as
the gap item: review the whole gap write handling process, refactor it and
fix handling of corner cases along the way.

Closes #7375

### CODE REVIEW

**Review by Korablev77 — APPROVED** [review]

File: src/box/memtx_tx.c [hunk 1]
```diff
@@ -1748,47 +1748,36 @@ memtx_tx_handle_gap_write(struct txn *txn, struct space *space,
...
+		bool is_eq = item->type == ITER_EQ || item->type == ITER_REQ;
+		bool is_e = item->type == ITER_LE || item->type == ITER_GE;
+		bool need_split = item->key == NULL ||
+				  (dir * cmp > 0 && !is_eq) ||
+				  (!is_full_key && cmp == 0 && (is_e || is_eq));
+		bool need_move = !need_split &&
+				 ((dir < 0 && cmp > 0) ||
+				  (cmp > 0 && item->type == ITER_EQ) ||
+				  (cmp == 0 && ((dir < 0 && is_full_key) ||
+						item->type == ITER_LT)));
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c:1773]: Neither did the existing code... Do you have any ideas on how to make it clear what's going on?

File: src/box/memtx_tx.c [hunk 2]
```diff
@@ -1748,47 +1748,36 @@ memtx_tx_handle_gap_write(struct txn *txn, struct space *space,
...
+							  item->part_count, oh,
+							  def);
 		}
-
-		if (is_split) {
+		int dir = iterator_direction(item->type);
+		bool is_full_key = item->part_count ==
+				   index->def->cmp_def->part_count;
+		bool is_eq = item->type == ITER_EQ || item->type == ITER_REQ;
+		bool is_e = item->type == ITER_LE || item->type == ITER_GE;
```
Comment (drewdzzz) [inline src/box/memtx_tx.c:1765]: It seems to me that you forgot about `ITER_ALL`.

**Review by drewdzzz — CHANGES_REQUESTED** [review]

File: src/box/memtx_tx.c (hunk 2 above)
Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c:1765]: Oh no, not again... ;c

Added a patch that denormalizes `ITER_ALL` to `ITER_GE` in TREE index, please take a look.

File: src/box/memtx_tx.c [hunk 3]
```diff
@@ -1234,6 +1234,7 @@ memtx_tx_story_full_unlink(struct memtx_story *story)
 			if (story->del_psn > 0 && link->in_index != NULL) {
 				struct index *index = link->in_index;
 				struct tuple *removed, *unused;
+				assert(link->older_story == NULL);
```
Comment (drewdzzz) [inline src/box/memtx_tx.c:1237]: Commit are mixed up, and something strange here

File: src/box/memtx_tx.c [hunk 4]
```diff
@@ -1297,6 +1298,9 @@ memtx_tx_story_gc_step()
 			    in_all_stories);
 	txm.traverse_all_stories = txm.traverse_all_stories->next;
 
+	if (flag)
+		flag += 0;
+
```
Comment (drewdzzz) [inline src/box/memtx_tx.c:1303]: And here :)

**Review by drewdzzz — CHANGES_REQUESTED** [review]

File: src/box/memtx_tx.c (hunk 4 above)
Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c]: Sorry again, fixed.

**Review by alyapunov — APPROVED** [review]

**Pull request merged**